include/pstl_offload/internal/usm_memory_replacement.h

```diff
@@ -0,0 +1,246 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) 2023 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _ONEDPL_PSTL_OFFLOAD_INTERNAL_USM_MEMORY_REPLACEMENT_H
+#define _ONEDPL_PSTL_OFFLOAD_INTERNAL_USM_MEMORY_REPLACEMENT_H
+
+#ifndef __SYCL_PSTL_OFFLOAD__
+#error "-fsycl-pstl-offload option should be passed to the compiler to use this header"
+#endif
+
+#include <atomic>
+#include <cstdlib>
+#include <cassert>
+#include <cerrno>
+
+#include <sycl/sycl.hpp>
+
+#include <oneapi/dpl/execution>
+
+#include "usm_memory_replacement_common.h"
+
+namespace __pstl_offload {
+
+static std::atomic<sycl::device*> __active_device = nullptr;
+
+static void __set_active_device(sycl::device* __new_active_device) {
+    __active_device.store(__new_active_device, std::memory_order_release);
+}
+
+class __offload_policy_holder_type {
```

Comment:
Why does this class, and whatever else related to policy substitution, reside in the memory replacement header, and not in a separate header?

Reply:
Actually, this class is not only a policy holder, but also an offload sycl::device object holder. And the device (as well as `active_device.store`) is required to make memory redirection.
Originally, this class was placed into the several header, but during the review, it was a comment that this separate header is included by `memory_replacement.h` and the algorithms part and never includes separately. And if inclusion of this header would be forgotten, memory replacement would not happen because active device would not be set.

Reply:
I do not understand all nuances and not completely convinced, but let's leave it as is for now and think again later.